sdk/program/src/message/sanitized.rs

```diff
@@ -0,0 +1,259 @@
+use {
+    crate::{
+        hash::Hash,
+        instruction::{CompiledInstruction, Instruction},
+        message::{MappedAddresses, MappedMessage, Message, MessageHeader},
+        pubkey::Pubkey,
+        sanitize::{Sanitize, SanitizeError},
+        serialize_utils::{append_slice, append_u16, append_u8},
+    },
+    std::convert::TryFrom,
+};
+
+/// Sanitized message of a transaction which includes a set of atomic
+/// instructions to be executed on-chain
+#[derive(Debug, Clone)]
+pub enum SanitizedMessage {
+    /// Sanitized legacy message
+    Legacy(Message),
+    /// Sanitized version #0 message with mapped addresses
+    V0(MappedMessage),
+}
+
+impl TryFrom<Message> for SanitizedMessage {
+    type Error = SanitizeError;
+    fn try_from(message: Message) -> Result<Self, Self::Error> {
+        message.sanitize()?;
+        Ok(Self::Legacy(message))
+    }
+}
+
+impl SanitizedMessage {
+    /// Return true if this message contains duplicate account keys
+    pub fn has_duplicates(&self) -> bool {
+        match self {
+            SanitizedMessage::Legacy(message) => message.has_duplicates(),
+            SanitizedMessage::V0(message) => message.has_duplicates(),
+        }
+    }
+
+    /// Message header which identifies the number of signer and writable or
+    /// readonly accounts
+    pub fn header(&self) -> &MessageHeader {
+        match self {
+            Self::Legacy(message) => &message.header,
+            Self::V0(mapped_msg) => &mapped_msg.message.header,
+        }
+    }
+
+    /// Returns a legacy message if this sanitized message wraps one
+    pub fn legacy_message(&self) -> Option<&Message> {
+        if let Self::Legacy(message) = &self {
+            Some(message)
+        } else {
+            None
+        }
+    }
+
+    /// Returns the fee payer for the transaction
+    pub fn fee_payer(&self) -> &Pubkey {
+        self.get_account_key(0)
```

Comment:
That is correct, but it's been awhile since we sanitized messages to ensure that programs can not be at index 0: https://github.com/solana-labs/solana/pull/9840/files#r418685785

Reply:
Ah right... I vaguely remember that conversation now.  We probably ought to remove the ambiguity at some point...